webview-ui/src/ConnectAcrToCluster/ConnectAcrToCluster.tsx

```diff
@@ -0,0 +1,280 @@
+                    labelGetter={(g) => g}
+                    onSelect={eventHandlers.onSetSelectedClusterResourceGroup}
+                />
+
+                <label htmlFor="cluster-input" className={styles.label}>
+                    Cluster
+                </label>
+                <ResourceSelector<Cluster>
+                    id="cluster-input"
+                    className={styles.control}
+                    resources={lazyClusters}
+                    selectedItem={state.selectedCluster}
+                    valueGetter={(c) => c.clusterName}
+                    labelGetter={(c) => c.clusterName}
+                    onSelect={eventHandlers.onSetSelectedCluster}
+                />
+
+                <label className={styles.label}>Authorization</label>
+                <div className={`${styles.control} ${styles.actionItemList}`}>
+                    <ActionItem {...getAcrAuthorizationActionItemProps()} />
+                </div>
+            </fieldset>
+        </>
+    );
+}
+
+type LocalData = {
+    lazySubscriptions: Lazy<Subscription[]>;
+    lazyAcrResourceGroups: Lazy<string[]>;
+    lazyClusterResourceGroups: Lazy<string[]>;
+    lazyAcrs: Lazy<Acr[]>;
+    lazyClusters: Lazy<Cluster[]>;
+    lazyAcrRoleState: Lazy<AcrRoleState>;
+};
+
+function prepareData(state: ConnectAcrToClusterState, updates: EventHandlerFunc[]): LocalData {
+    const lazySubscriptions = ensureSubscriptionsLoaded(state.azureReferenceData, updates);
+    const lazySubscriptionAcrs = ensureAcrsLoaded(state.azureReferenceData, state.selectedSubscription, updates);
+    const lazyAcrResourceGroups = lazyMap(lazySubscriptionAcrs, (acrs) => distinct(acrs.map((a) => a.resourceGroup)));
+    const lazySubscriptionClusters = ensureClustersLoaded(
+        state.azureReferenceData,
+        state.selectedSubscription,
+        updates,
+    );
+    const lazyClusterResourceGroups = lazyMap(lazySubscriptionClusters, (clusters) =>
+        distinct(clusters.map((c) => c.resourceGroup)),
+    );
+    const lazyAcrs = lazyMap(lazySubscriptionAcrs, (acrs) =>
+        acrs.filter((a) => a.resourceGroup === state.selectedAcrResourceGroup),
+    );
+    const lazyClusters = lazyMap(lazySubscriptionClusters, (clusters) =>
+        clusters.filter((c) => c.resourceGroup === state.selectedClusterResourceGroup),
+    );
+    const lazyAcrRoleState = ensureAcrRoleAssignmentLoaded(
+        state.azureReferenceData,
+        state.selectedSubscription,
+        state.selectedAcr,
+        state.selectedCluster,
+        updates,
+    );
+
+    return {
+        lazySubscriptions,
+        lazyAcrResourceGroups,
+        lazyClusterResourceGroups,
+        lazyAcrs,
+        lazyClusters,
+        lazyAcrRoleState,
+    };
+}
+
+function makeFixAction(icon: IconDefinition, name: string, action: (() => void) | null): FixAction {
+    return {
+        icon,
+        name,
+        action: action ? action : () => {},
+        canPerformAction: action !== null,
```

Comment:
instead of going by action !== null, can we pass in hasAcrPull boolean variable so that buttons can be enabled? so that we can avoid passing null values.

Reply:
Unfortunately, no. There will be times when we don't know the value of `hasAcrPull`. The `action` parameter is intentionally nullable to indicate _either_ the action to perform _or_ `null` meaning action not applicable.

Reply:
hasAcrPull is boolean variable true | false , passing null is not good practice. canPerformAction: hasAcrpull === true? something like that. this is kind suggestion. if this doesn't work out you have already added comments. it should be good to go. Thank you.